homeassistant/helpers/service.py

```diff
@@ -360,3 +362,40 @@ def async_register_admin_service(
     hass.services.async_register(
         domain, service, admin_handler, schema
     )
+
+
+@bind_hass
+@ha.callback
+def verify_domain_control(hass: HomeAssistantType, domain: str) -> Callable:
+    """Ensure permission to access any entity under domain in service call."""
+    def decorator(service: Callable) -> Callable:
+        """Decorate."""
+        async def check_permissions(call):
+            """Check user permission and raise before call if unauthorized."""
+            if not call.context.user_id:
+                return hass.async_add_job(service, call)
```

Comment:
this does not what you think it does. You should await it.

Reply:
```suggestion
                return await hass.async_add_job(service, call)
```

Reply:
Actually, we should limit this decorator to just async functions and just await it here.

Raise error if [`asyncio.iscoroutinefunction`](https://docs.python.org/3/library/asyncio-task.html#asyncio.iscoroutinefunction) returns false for service handler

Reply:
Will callbacks not work? What about checking that the service handler is neither a callback nor a coroutine?

Reply:
Doesn't that limit us unnecessarily? What if the service handler is a callback?

```
    @ha.callback
    def mock_service_log(call):
        """Define a protected service."""
        calls.append(call)
```

Reply:
that's just a test. There is no real world scenario where it's not async method because it is supposed to interact with things, which requires I/O